torch/_dynamo/optimizations/distributed.py

```diff
@@ -236,42 +247,69 @@ def forward(self, *args):
-
+                for sn in fake_submod.graph.nodes:
+                    if sn.op == "output":
+                        if not isinstance(sn.args[0], tuple):
+                            unwrap_singleton_tuple = True
+                            sn.args = (sn.args,)
+                real_mod.recompile()
+                fake_submod.recompile()
                 wrapper = WrapperModule(
-                    self.compiler(submod, args),
+                    self.compiler(real_mod, args),
+                    self.compiler(fake_submod, args),
                     unwrap_singleton_tuple,
                 )
                 return wrapper
 
             def run_node(self, n: Node) -> Any:
                 with fx_traceback.append_stack_trace(n.stack_trace):
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
+                    fake_mode = None
+                    for tensor in args:
+                        if isinstance(tensor, torch._subclasses.FakeTensor):
+                            fake_mode = tensor.fake_mode
+                            break
+                    new_args = []
+                    if fake_mode:
+                        for arg in args:
+                            if isinstance(arg, torch.Tensor) and not isinstance(arg, torch._subclasses.FakeTensor):
+                                new_args.append(fake_mode.from_tensor(arg))
+                            else:
+                                new_args.append(arg)
+                    else:
+                        new_args = args
+
                     log.debug(f"run_node {n.op}, {n.target} got args {args_str(args)}")
                     assert isinstance(args, tuple)
                     assert isinstance(kwargs, dict)
 
                     # modify the currently running FX graph
                     # maybe this isn't sound in general, but only changing the target of a node might be ok?
                     if n.op == "call_module":
-                        submod = self.fetch_attr(n.target)
-                        log.debug(f"\n---{n.target} graph---\n" + str(submod.graph))
-                        compiled_submod = self.compile_submod(submod, args, kwargs)
+                        real_mod = self.fetch_attr(n.target)
+                        if fake_mode:
+                            from ..utils import deepcopy_to_fake_tensor
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
+                        else:
+                            fake_submod = real_mod
+                            pass
+                        log.debug(f"\n---{n.target} graph---\n" + str(fake_submod.graph))
+                        compiled_submod = self.compile_submod(fake_submod, real_mod, new_args, kwargs)
                         self.module.delete_submodule(n.target)
                         n.target = "compiled_" + n.target
                         self.module.add_submodule(n.target, compiled_submod)
```

Comment:
manually call (and return) `fake_submod(*args, **kwargs)` (to continue propagating fake tensors). DO NOT modify the input FX graph, this is no longer necessary